libs/wingc/src/ast.rs

```diff
@@ -409,16 +409,23 @@ impl BinaryOperator {
 }
 
 #[derive(Debug)]
-pub enum Reference {
-	Identifier(Symbol),
+pub struct Reference {
+	pub kind: ReferenceKind,
+	pub evaluated_type: RefCell<Option<TypeRef>>,
+	pub span: WingSpan,
+}
+
+#[derive(Debug)]
+pub enum ReferenceKind {
+	Identifier { symbol: Symbol },
```

Comment:
This decoupling looks good, though I'm wondering doesn't this mean we can end up with structures like this?

```rust
Expr {
  kind: ExprKind::Reference(Reference {
    kind: ReferenceKind::Identifier { .. }
    evaluated_type: // A
  }),
  evaluated_type: // B
}
```

Is there is a reason we need the `evaluated_type` data in both places given that they will always have the same value, in this case anyways? (disclaimer, I don't fully grasp all of the functions `Reference` serves in the compiler)

(random thought - something about this reminds me a little bit of lvalues vs rvalues https://eli.thegreenplace.net/2011/12/15/understanding-lvalues-and-rvalues-in-c-and-c)

Reply:
Yeah that situation is possible and in fact would probably happen often. It occurs any time an expression is just a reference, e.g. `x + a.b` where both `x` and `a.b` are such expressions.

The reason I added the additional evaluated_type is primarily for the case of call expressions like `a.b()`. In this case, the evaluated type of the whole expression is not the same as the Reference `a.b` and there is no way to find the type of the reference unless you're inside the TypeChecker.

There is definitely a smell of lvalues and rvalues here. References are our only lvalues currently so they are unique. `Expr` are our rvalues but they are allowed to wrap References. Not sure if it affects any changes here other than some reasoning why Reference can get some special treatment.